src/api/submission/validations.py

```diff
@@ -1075,56 +1075,58 @@ def get_observations(cls, data, key=None):
+        has_missing_values = False
+        for ob in obs:
+            pct_hard = ob.get("percent_hard")
+            pct_soft = ob.get("percent_soft")
+            pct_algae = ob.get("percent_algae")
 
-        if value < 0:
-            return ERROR, self.LESS_EQUAL_0_MSG.format(msg_param)
-        elif value > 100:
-            return ERROR, self.GREATER_100_MSG.format(msg_param)
-        return OK, ""
+            pct_values = [pct_algae, pct_hard, pct_soft]
 
-    def _validate_percent_values(self, values, validation, msg_param="value"):
-        for val in values:
-            result, msg = self._check_percent_value(val, msg_param)
-            if result == ERROR:
-                return self.error(self.identifier, msg, validation=validation)
-        return OK
+            if any(pv < 0 or pv > 100 for pv in pct_values if pv):
+                return self.error(
+                    self.identifier,
+                    self.GREATER_100_MSG.format("value"),
+                    validation="validate_percent_values"
+                )
 
-    def _validate_percent_value_total(self, values, validation):
-        total = 0
-        for v in values:
-            try:
-                total += float(v)
-            except (TypeError, ValueError):
-                pass
-        return self._validate_percent_values([total], validation, msg_param="total")
+            if safe_sum(pct_algae, pct_hard, pct_soft) > 100:
+                return self.error(
+                    self.identifier,
+                    self.GREATER_100_MSG.format("total"),
+                    validation="validate_percent_values"
+                )
 
-    def validate_percent_values(self):
+            if None in pct_values:
+                has_missing_values = True
+                continue
+
+        if has_missing_values:
+            return self.warning(self.identifier, self.VALUE_NOT_SET)
+
+        return self.ok(self.identifier)
+
+    def validate_quadrat_numbers(self):
         obs = self.get_observations(self.data)
+        quadrat_nums = []
         for ob in obs:
-            percent_hard = ob.get("percent_hard")
-            percent_soft = ob.get("percent_soft")
-            percent_algae = ob.get("percent_algae")
-            values = [percent_hard, percent_soft, percent_algae]
-            if (
-                self._validate_percent_values(values, "validate_percent_values")
-                == ERROR
-            ):
-                return ERROR
+            quadrat_num = ob.get("quadrat_number")
+            if quadrat_num is not None and quadrat_num <= 0:
+                return self.error(self.identifier, self.LESS_QUADRAT_NUMBER)
 
-            if (
-                self._validate_percent_value_total(values, "validate_percent_values")
-                == ERROR
-            ):
-                return ERROR
+            quadrat_nums.append(quadrat_num)
+
+        if len(quadrat_nums) != len(set(quadrat_nums)):
```

Comment:
This tests that the number of quadrat numbers is the same -- but couldn't they be the same and still have a duplicate? I'm probably not seeing the whole logic.

Reply:
it's checking the length of the list versus the length of the set of quadrat_nums ?